### PR TITLE
handle multiple packages providing same files sanely

### PR DESCRIPTION
**Pull request by deitch**
This is an issue for dependencies that are not names but in the "provides" list. Previously, it just took the last one, which could be different each time.

Here, it improves, if not completes it:

1. If the same package we are using provides the dependency, then nothing to add
2. If another package that provides the dependency has the same origin as our package, use that
3. Otherwise, use the first one on the list

The last step still isn't great, but it is an improvement as we add the first 2 steps. 

@kaniini pointed out that the apk-tools logic is:

> if a package is already enqueued to be installed, and it provides something, then that package will be considered first for the provided dependency

We should do that, but in a later PR.

### CODE REVIEW

**Review by jdolitsky — APPROVED** [review]

**Review by jdolitsky — CHANGES_REQUESTED** [review]
waiting on ci

File: pkg/apk/impl/repo.go [hunk 1]
```diff
@@ -285,30 +287,34 @@ func (p *PkgResolver) GetPackageWithDependencies(pkgName string) (pkg *repositor
...
+		// get the one that most matches what was requested
+		versions := make([]string, 0, 10)
+		for version := range pkgsWithVersions {
+			versions = append(versions, version)
+		}
+		targetVersion := getBestVersion(versions, version, compare)
+		if targetVersion == "" {
+			return nil, nil, nil, fmt.Errorf("could not find package %s in indexes: %w", pkgName, err)
+		}
+		pkg = pkgsWithVersions[targetVersion]
```
Comment (jdolitsky) [inline pkg/apk/impl/repo.go:301]: nil deref here?

File: pkg/apk/impl/repo.go [hunk 2]
```diff
@@ -378,33 +384,45 @@ func (p *PkgResolver) getPackageDependencies(pkg *repository.RepositoryPackage,
...
+			// get the one that most matches what was requested
+			var versions []string
+			for version := range depPkgWithVersions {
+				versions = append(versions, version)
+			}
+			targetVersion := getBestVersion(versions, version, compare)
+			if targetVersion == "" {
+				return nil, nil, fmt.Errorf("could not find package %s in indexes: %w", dep, err)
+			}
+			depPkg = depPkgWithVersions[targetVersion]
```
Comment (jdolitsky) [inline pkg/apk/impl/repo.go:398]: nil dereference here?

**Comment by deitch** [comment]
Ah, now it is much happier

**Review by jdolitsky — APPROVED** [review]

**Pull request merged**